target/linux/ath79/files/drivers/mfd/rb4xx-cpld.c

```diff
@@ -0,0 +1,197 @@
+ *
+ * The CPLD supports SPI two-wire mode, in which two bits are transferred per
+ * SPI clock cycle. The second bit is transmitted with the SoC's CS2 pin.
+ *
+ * The CPLD also acts as a GPIO expander.
+ *
+ * Copyright (C) 2008-2011 Gabor Juhos <juhosg@openwrt.org>
+ * Copyright (C) 2008 Imre Kaloz <kaloz@openwrt.org>
+ * Copyright (C) 2015 Bert Vermeulen <bert@biot.com>
+ * Copyright (C) 2020 Christopher Hill <ch6574@gmail.com>
+ *
+ * This file was based on the driver for Linux 2.6.22 published by
+ * MikroTik for their RouterBoard 4xx series devices.
+*/
+#include <linux/mfd/core.h>
+#include <linux/spi/spi.h>
+#include <linux/module.h>
+#include <linux/of_platform.h>
+
+#include <mfd/rb4xx-cpld.h>
+
+/* CPLD commands */
+#define CPLD_CMD_WRITE_NAND	0x08 /* send cmd, n x send data, send idle */
+#define CPLD_CMD_WRITE_CFG	0x09 /* send cmd, n x send cfg */
+#define CPLD_CMD_READ_NAND	0x0a /* send cmd, send idle, n x read data */
+#define CPLD_CMD_READ_FAST	0x0b /* send cmd, 4 x idle, n x read data */
+#define CPLD_CMD_GPIO8_HIGH	0x0c /* send cmd */
+#define CPLD_CMD_GPIO8_LOW	0x0d /* send cmd */
+
+static int rb4xx_cpld_write_nand(struct rb4xx_cpld *cpld, const void *tx_buf,
+			    unsigned int len)
+{
+	struct spi_message m;
+	static const u8 cmd = CPLD_CMD_WRITE_NAND;
+	struct spi_transfer t[3] = {
+		{
+			.tx_buf = &cmd,
+			.len = sizeof(cmd),
+		}, {
+			.tx_buf = tx_buf,
+			.len = len,
+			.tx_nbits = SPI_NBITS_DUAL,
+		}, {
+			.len = 1,
+			.tx_nbits = SPI_NBITS_DUAL,
+		},
+	};
+
+	spi_message_init(&m);
+	spi_message_add_tail(&t[0], &m);
+	spi_message_add_tail(&t[1], &m);
+	spi_message_add_tail(&t[2], &m);
+	return spi_sync(cpld->spi, &m);
+}
+
+static int rb4xx_cpld_read_nand(struct rb4xx_cpld *cpld, void *rx_buf,
+				unsigned int len)
+{
+	struct spi_message m;
+	static const u8 cmd[2] = {
+		CPLD_CMD_READ_NAND, 0
+	};
+	struct spi_transfer t[2] = {
+		{
+			.tx_buf = &cmd,
+			.len = sizeof(cmd),
+		}, {
+			.rx_buf = rx_buf,
+			.len = len,
+		},
+	};
+
+	spi_message_init(&m);
+	spi_message_add_tail(&t[0], &m);
+	spi_message_add_tail(&t[1], &m);
+	return spi_sync(cpld->spi, &m);
+}
+
+static int rb4xx_cpld_cmd(struct rb4xx_cpld *cpld, const void *tx_buf,
+			  size_t len)
+{
+	struct spi_message m;
+	struct spi_transfer t = {
+		.tx_buf = tx_buf,
+		.len = len,
+	};
+
+	spi_message_init(&m);
+	spi_message_add_tail(&t, &m);
+	return spi_sync(cpld->spi, &m);
+}
+
+static int rb4xx_cpld_set_gpios(struct rb4xx_cpld *cpld, u16 values)
+{
+	static const u8 gpio8_low  = CPLD_CMD_GPIO8_LOW;
+	static const u8 gpio8_high = CPLD_CMD_GPIO8_HIGH;
+	u8 low8 = (u8) values;
+	u8 cmd[2] = { CPLD_CMD_WRITE_CFG, low8 };
+	int ret;
+
+	/* GPIOs 0-7 can be sent direct to CPLD */
+	if (low8 != (u8) cpld->values) {
+		ret = rb4xx_cpld_cmd(cpld, &cmd, ARRAY_SIZE(cmd));
+		if (unlikely(ret)) {
+			return ret;
+		}
+	}
+
+	/* GPIO 8 is a special case */
+	if ((values & BIT(8)) != (cpld->values & BIT(8))) {
```

Comment:
I think this is doubly wrong: if I read this function correctly, I understand that the cpld commands can really only set one gpio at a time (otherwise setting gpios 0, 2 and 3 together would equate turning LED5 on); if that's really the case then 1) the bug I mention above will trigger, 2) this function should probably be renamed to avoid plural and 3) it can be rewritten in a much more efficient way. I'm signing off for now but I'll revisit tomorrow if I'm not completely besides myself :)

Reply:
In theory the cpld allows setting multiple gpios at the same time (it just takes the bit field and applies it), however the gpio interface prevents this via only allowing one offset in the `set` call at a time.

So I have the first 8 bits in `values` as gpio 0-7, and we can toggle one bit on/off at a time. We only send the SPI command if we have a change in that first byte.

The higher 8 bits in `values` are only for gpio 8, and again we can toggle that bit on/off too. If that changes we send a slightly different SPI command.

I've tested this out by doing `echo 1 > /sys/class/leds/mikrotik\:green\:led5/brightness` and watching the 5 leds turn on and off individually.

N.B. the 5th led complicates this whole thing, and I now kinda wish they didn't include it! :smiley: